File: performance-comparison/benchmarks/haskell-project/src/FileRead.hs
-- file_read -- write an N-byte 0xCD file, read back in 4 KiB chunks.
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
import System.IO
import System.Directory (removeFile)
import qualified Data.ByteString as BS
main :: IO ()
main = do
  args <- getArgs
  let n = case args of (a:_) -> read a; _ -> 1048576 :: Int
      path = "/tmp/bench_io_read_hs.bin"
      chunk = BS.replicate 4096 0xCD
  h <- openBinaryFile path WriteMode
  let put !rem_
        | rem_ <= 0 = return ()
        | otherwise = BS.hPut h (BS.take (min 4096 rem_) chunk) >> put (rem_ - 4096)
  put n
  hClose h
  r <- openBinaryFile path ReadMode
  let get !total = do
        b <- BS.hGet r 4096
        if BS.null b then return total else get (total + BS.length b)
  total <- get 0
  hClose r
  removeFile path
  print total

File: performance-comparison/benchmarks/rust-workspace/src/bin/file_read.rs
// file_read -- write n bytes (0xCD) then read them back in 4 KiB chunks,
// print bytes read.
use std::env;
use std::fs;
use std::io::{Read, Write};

const CHUNK: usize = 4096;

fn file_read_bench(n: i64) -> i64 {
    let n = n as usize;
    let path = "/tmp/bench_io_read_rs.bin";
    let buf = [0xCDu8; CHUNK];
    {
        let mut f = fs::File::create(path).unwrap();
        let mut rem = n;
        while rem > 0 {
            let chunk = std::cmp::min(CHUNK, rem);
            f.write_all(&buf[..chunk]).unwrap();
            rem -= chunk;
        }
    }
    let mut f = fs::File::open(path).unwrap();
    let mut rbuf = [0u8; CHUNK];
    let mut total: i64 = 0;
    loop {
        let nr = f.read(&mut rbuf).unwrap();
        if nr == 0 {
            break;
        }
        total += nr as i64;
    }
    drop(f);
    let _ = fs::remove_file(path);
    total
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1048576);
    println!("{}", file_read_bench(n));
}
